src/vos/tests/vts_wal.c

```diff
@@ -207,17 +243,676 @@ wal_tst_01(void **state)
+
+	/* Fetch/verify KVs */
+	for (oidx = 0; oidx < obj_nr; oidx++) {
+		arg->oid = oids[oidx];
+
+		for (didx = 0; didx < dkey_nr; didx++) {
+			if (!overwrite) {
+				wal_dkey_gen(&dkey, arg);
+				rex.rx_idx = hash_key(&dkey,
+					is_daos_obj_type_set(arg->otype, DAOS_OT_DKEY_UINT64));
+			}
+
+			for (aidx = 0; aidx < v_nr; aidx++) {
+				wal_print_buf(update_buf, aidx + v_nr * (didx + dkey_nr * oidx));
+				if (!overwrite)
+					wal_akey_gen(&iod.iod_name, arg);
+
+				iod.iod_size = DAOS_REC_ANY;
+
+				rc = io_test_obj_fetch(arg, ep++, 0, &dkey, &iod, &sgl, true);
+				assert_rc_equal(rc, 0);
+
+				if (arg->ta_flags & TF_REC_EXT)
+					assert_int_equal(iod.iod_size, UPDATE_REC_SIZE);
+				else
+					assert_int_equal(iod.iod_size, UPDATE_BUF_SIZE);
+				assert_memory_equal(update_buf, fetch_buf, UPDATE_BUF_SIZE);
+			}
+		}
+	}
+}
+
+static void
+wal_io_multiple_objects(void **state)
+{
+	struct io_test_args	*arg = *state;
+	daos_epoch_t		 epoch;
+	int i;
+
+	num_keys = WAL_OBJ_KEYS;
+
+	for (i = 0; io_test_flags[i].tf_str != NULL; i++) {
+		print_message("\t%d) multiple objects update (%s) test\n",
+			      i, io_test_flags[i].tf_str);
+
+		epoch = gen_rand_epoch();
+		arg->ta_flags = io_test_flags[i].tf_bits;
+
+		/* Update KVs in num_keys objects, refill pool and fetch/verify all values */
+		wal_objs_update_and_fetch(arg, epoch);
+	}
+}
+
+static void
+wal_io_multiple_objects_ovwr(void **state)
+{
+	struct io_test_args	*arg = *state;
+	daos_epoch_t		 epoch;
+	int i;
+
+	num_keys = WAL_OBJ_KEYS;
+
+	for (i = 0; io_test_flags[i].tf_str != NULL; i++) {
+		print_message("\t%d) multiple objects overwrite (%s) test\n",
+			      i, io_test_flags[i].tf_str);
+
+		epoch = gen_rand_epoch();
+		arg->ta_flags = io_test_flags[i].tf_bits;
+		arg->ta_flags |= TF_OVERWRITE;
+
+		/*
+		 * Update same key value in num_keys objects,
+		 * refill pool and fetch/verify the values
+		 **/
+		wal_objs_update_and_fetch(arg, epoch);
+	}
+}
+
 static const struct CMUnitTest wal_tests[] = {
 	{ "WAL01: Basic pool operations",
 	  wal_tst_01, NULL, NULL },
 };
 
+static const struct CMUnitTest wal_io_tests[] = {
+	{ "WAL10: Update/fetch/verify test",
+	  wal_io_multiple_refills, NULL, NULL },
+	{ "WAL11: 10K update/fetch/verify test",
+	  wal_io_multiple_updates, NULL, NULL },
+	{ "WAL12: Objects Update(overwrite)/fetch test",
+	  wal_io_multiple_objects_ovwr, NULL, NULL },
+	{ "WAL13: Objects Update/fetch test",
+	  wal_io_multiple_objects, NULL, NULL },
+};
+
+static const struct CMUnitTest wal_io_int_tests[] = {
+	{ "WAL20: Key query punch with subsequent update",
+	  wal_io_query_key_punch_update, NULL, NULL },
+};
+
 int
 run_wal_tests(const char *cfg)
 {
-	char	test_name[DTS_CFG_MAX];
-
-	dts_create_config(test_name, "WAL Tests %s", cfg);
-	return cmocka_run_group_tests_name(test_name, wal_tests, setup_wal_test,
+	char		 test_name[DTS_CFG_MAX];
+	const char	*akey = "hashed";
+	const char	*dkey = "hashed";
+	int		 i, rc;
+
```

Comment:
All tests should be skipped when DAOS_MD_ON_SSD isn't enabled.

Reply:
I agree. I'll update.